### DISCUSSION TITLE
Tool for markdown-it to escape the necessary characters in markdown so I can try autorender

### DISCUSSION THREAD

**Discussion by universemaster** [KaTeX/KaTeX#3362] [Q&A] [answered]
Hi,

I usually use server-side rendering with [vscode-markdown-katex](https://github.com/mjbvz/vscode-markdown-katex).

I want to try in-browser rendering with [autorender](https://katex.org/docs/autorender.html), but I still want to use my markdown processor for it's other work.

Do you know of a tool/plug-in I can use to do the [escaping](https://katex.org/docs/api.html) required?  I use [markdown-it](https://www.npmjs.com/package/markdown-it).

I could of course do write the escaping code myself, but I suspect there are lots of edge-cases that I will be chasing for weeks.

I'm asking for my own reasons, but also think it would be good to detail this on the appropriate pages of www.katex.org.

**Comment by edemaine** [comment]
May I ask why you want to use autorender?

The general recommendation is to use a markdown plugin that interprets all the math-entering symbols and calls KaTeX while rendering the Markdown, instead of escaping them and then calling autorender.  For markdown-it, I've used [markdown-it-texmath](https://github.com/goessner/markdown-it-texmath) with good results.

I agree that this should probably be added to https://katex.org/docs/libs.html

**Comment by universemaster** [comment]
Your questioning makes me think maybe it's not the way to go for better performance.

My reason for the interest is that when I enable `content-visibility: auto` (see other discussion https://github.com/KaTeX/KaTeX/discussions/3357 ) for browsers that support it, my limiting performance factor is **Parsing HTML and CSS**. So I was trying to send the raw KaTeX source for my equations, rather than the pre-rendered DOM tree.

---

Let me use this discussion thread to give a little more background regarding where I am with optimising performance.

Notice (on the screenshot below) that on desktop (and sadly desktop only), I can get a decent lighthouse score despite the massive (understatement!) DOM.

Notice also, that the most main-thread work (relating to total blocking time) is parsing HTML and CSS.

Disappointingly, mobile browsers and those that don't support `content-visibility` give a very different story (also less good when I switch to a remote server - or use a cloudflare tunnel which frustratingly doesn't support brotli from origin to cloudflare!  My gzipped html is double the size of my brotli html).

These scores are **with** all the optimisations in the other threads and are for {output: html}, desktop chrome browser, local server (with brotli max level 11) and `content-visibility: auto` enabled.

For the stats on my other threads, I commented out `content-visibility` from my stylesheet - since it doesn't work well on mobile, or safari on any device.

![image](https://user-images.githubusercontent.com/750365/138900602-adf9851e-2034-4ed3-9c01-b8dc76f0f487.png)

![image](https://user-images.githubusercontent.com/750365/138901182-191dac63-495c-470c-b249-ce5bf9c3aa08.png)

**Comment by universemaster** [comment]
And mobile with otherwise the same settings (including content-visibility: auto - so it will stutter when I scroll).

![image](https://user-images.githubusercontent.com/750365/138906570-54b80555-7a62-43d5-b4e9-cd410553f6ce.png)

![image](https://user-images.githubusercontent.com/750365/138906704-4fb5bb17-7f2c-4957-8e5e-a2ad20eb6d04.png)

**Comment by edemaine** [comment, marked as answer]
OK, makes sense now; you want to render only when the equation becomes visible.

The timing numbers you give lack some context: could you summarize how big the DOM is, how many equations, average DOM size per equation, etc.?  This would help put the parse times into perspective.

I think the way to do this would be to use a plugin like `markdown-it-texmath`, but give it a fake KaTeX engine like so (untested code):
```js
const markdown = require('markdown-it')({html: true})
.use(require('markdown-it-texmath'), {
  engine: {
    renderToString: (tex, options) => `<div class="math-todo" data-display="${options.displayMode}">${tex}</div>`
  }
})
```
Then you could write some JavaScript to build an [`IntersectionObserver`](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API) to detect when these become visible and call KaTeX to render them, and possibly reset them when they're no longer visible.  This on-demand KaTeX rendering code would probably be an interesting package to write, either stand-alone or for `contrib` here.

**Comment by universemaster** [comment]
But if I comment out all `content-visibility:auto` and `contain:` from my stylesheet and keep everything else the same, **all the katex rendering needs to happen at page load** (hence the issues I raised) and we get an almost laughable difference - here is **desktop**, which lighthouse timed-out after it took too long (sorry for the wait for this next post) - notice it didn't load the entire DOM!  So it's actually even worse than the 45 seconds of main-thread!

(I shall not bother giving you the mobile version!)

![image](https://user-images.githubusercontent.com/750365/138910821-5d8bc60c-073e-421a-b664-00808fadb5e5.png)

![image](https://user-images.githubusercontent.com/750365/138911217-f3e7ef48-6f31-4444-9122-1de6ad109589.png)

**Comment by universemaster** [comment]
Sorry, I should have highlighted that the true DOM size is 562,872 elements - the first item under diagnostics (and the test without content-visibility didn't even load all of that).

There are 16,000 "equations", which I put in quotes become some are just $x$.  "Almost all" of those elements are from KaTeX.

**Comment by mbourne** [comment]
@universemaster Can you please provide a link to your 16,000-"equation" page so I can run some experiments? Thanks.

**Reply by universemaster** [reply]
Sure, I've created a fresh repository, and added you as a collaborator.  Feel free to share with other KaTeX people, but I'm not ready for it to be 100% live on the public internet yet.

You'll be interested in Minified-1A-All-Courses.html.  I've added the brotli files too.

**Reply by universemaster** [reply]
I've added an un-minimised version of the main stylesheet `preview.css` with comments.

**Reply by edemaine** [reply]
I'd be curious to see it too.

**Reply by universemaster** [reply]
I was waiting for that - added!

**Reply by kelvindi** [reply]
@universemaster Can you please add me too? I also have same question and want to explore

**Reply by universemaster** [reply]
Sure, added.

**Comment by universemaster** [comment]
Relating to this: If you scramble the non-KaTeX text (and maybe the "text" bits inside the KaTeX too), you could this html file for testing in the official KaTeX repository.

By `scramble` I mean replace each character with a random other character.

Relates to https://github.com/KaTeX/KaTeX/issues/2338

**Reply by universemaster** [reply]
Oh, you'd need to scramble the original markdown - this is getting complex!
